### PR TITLE
UpTime

### PR DESCRIPTION
**Pull request by vgomezASC**
These changes are what the tutorial told me to do for the mini feature
Up Time.

### CODE REVIEW

**Comment by lepidopterane-atsmith** [comment]
Hello, I'll be posting my code review here. The way Yuhang & I have been doing this is we post a stream of comments on the pull we're reviewing, and when the author of reviewed code finds this, they discuss the review w/the reviewer (pg16 channel or DM works fine for me). 

My review consists of an intro comment, a universal comment if necessary, and one comment for each class I have feedback about. I will **bold** questions I have, and you should answer them via Slack as you read the review. Let's begin.

**Comment by lepidopterane-atsmith** [comment]
_Universal Comment_
**Tell me about the tests you ran on your feature.**

**Comment by lepidopterane-atsmith** [comment]
_Chat.java_
Lines 45 and 48 ask to be replaced with specific code in their space. When communicating errors, a System.out.println(); does the trick.

**Comment by lepidopterane-atsmith** [comment]
_View.java_
Lines 148 and 152 ask to be replaced with specific code in their space.

**Comment by lepidopterane-atsmith** [comment]
_Server_Info.java_
- First off, it’s ServerInfo.java. I misunderstood and told you Yuhang’s working on that but both of you have to work on different versions that will eventually form Voltron. I’ll talk more about this later tonight
- **What's the deal with everything beyond line 9?**

**Comment by lepidopterane-atsmith** [comment]
_Server.java_
- Line 95 should not terminate in …, and compiling Server.java would’ve caught that error. You should look over what the tutorial says for that chunk to gain better insight on what that chunk does.
- This goes for the analogous chunks with similar endings in lines 112, 130, 146, 162, 182, and 199.
- Again this is because I gave a vague answer but line 311 should be moved to the top or some other logical place based on where the variable info is used in that program. It may have to be copy/pasted more than once, idk.

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -37,6 +37,18 @@
...
   public Chat(Context context) {
     this.panels.push(createRootPanel(context));
+      panel.register(“info”, new Panel.Command() {
+        @Override
+        public void invoke(Scanner args) {
+            final ServerInfo info = context.getInfo();
+            if (info == null) {
+                System.out.println("The server did not send us a valid info object")
+                } else {
+                // Print the server info to the user in a pretty way
```
Comment (aslo) [inline src/codeu/chat/client/commandline/Chat.java]: You need to add some of your own code here that displays the uptime information to the user. Take a look at @ComarPers922 's pull request https://github.com/vgomezASC/CodeU-Summer-2017/pull/4 that adds a similar feature using the `ServerInfo` object.

File: src/codeu/chat/server/Server.java [hunk 2]
```diff
@@ -280,6 +308,7 @@ public void run() {
                     relay.pack(user.id, user.name, user.creation),
                     relay.pack(conversation.id, conversation.title, conversation.creation),
                     relay.pack(message.id, message.content, message.creation));
+  private static final ServerInfo info = new ServerInfo();
```
Comment (aslo) [inline src/codeu/chat/server/Server.java]: This should be added close to the top of this file, near the other similar declarations of member variables, maybe around line 55 or so?

File: src/codeu/chat/server/Server.java [hunk 3]
```diff
@@ -89,9 +89,13 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
             author,
             conversation,
             message.id));
+     if (type == NetworkCode.SERVER_INFO_REQUEST) {
```
Comment (aslo) [inline src/codeu/chat/server/Server.java]: This isn't quite right, but it looks like the tutorial is a bit out of date. You should add a new command in this class  - your code should look something like 
```
     this.commands.put(NetworkCode. SERVER_INFO_REQUEST,  new Command() {
        @Override
       public void onMessage(InputStream in, OutputStream out) throws IOException {
        Serializers.INTEGER.write(out, NetworkCode.SERVER_INFO_RESPONSE);
        Uuid.SERIALIZER.write(out, info.version); 
        }
    }
```

File: src/codeu/chat/server/Server.java [hunk 4]
```diff
@@ -102,6 +106,10 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
 
         Serializers.INTEGER.write(out, NetworkCode.NEW_USER_RESPONSE);
         Serializers.nullable(User.SERIALIZER).write(out, user);
+      if (type == NetworkCode.SERVER_INFO_REQUEST) {
```
Comment (aslo) [inline src/codeu/chat/server/Server.java]: See my above comment - you only need to make that change once, so you can probably delete this section.

File: src/codeu/chat/server/Server.java [hunk 5]
```diff
@@ -116,6 +124,10 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
 
         Serializers.INTEGER.write(out, NetworkCode.NEW_CONVERSATION_RESPONSE);
         Serializers.nullable(ConversationHeader.SERIALIZER).write(out, conversation);
+      if (type == NetworkCode.SERVER_INFO_REQUEST) {
```
Comment (aslo) [inline src/codeu/chat/server/Server.java]: See my above comment - you only need to make that change once, so you can probably delete this section.

File: src/codeu/chat/server/Server.java [hunk 6]
```diff
@@ -128,6 +140,10 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
 
         Serializers.INTEGER.write(out, NetworkCode.GET_USERS_RESPONSE);
         Serializers.collection(User.SERIALIZER).write(out, users);
+      if (type == NetworkCode.SERVER_INFO_REQUEST) {
```
Comment (aslo) [inline src/codeu/chat/server/Server.java]: See my above comment - you only need to make that change once, so you can probably delete this section.

File: src/codeu/chat/server/Server.java [hunk 7]
```diff
@@ -140,6 +156,10 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
 
         Serializers.INTEGER.write(out, NetworkCode.GET_ALL_CONVERSATIONS_RESPONSE);
         Serializers.collection(ConversationHeader.SERIALIZER).write(out, conversations);
+      if (type == NetworkCode.SERVER_INFO_REQUEST) {
```
Comment (aslo) [inline src/codeu/chat/server/Server.java]: See my above comment - you only need to make that change once, so you can probably delete this section.

File: src/codeu/chat/server/Server.java [hunk 8]
```diff
@@ -156,6 +176,10 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
 
         Serializers.INTEGER.write(out, NetworkCode.GET_CONVERSATIONS_BY_ID_RESPONSE);
         Serializers.collection(ConversationPayload.SERIALIZER).write(out, conversations);
+      if (type == NetworkCode.SERVER_INFO_REQUEST) {
```
Comment (aslo) [inline src/codeu/chat/server/Server.java]: See my above comment - you only need to make that change once, so you can probably delete this section.

File: src/codeu/chat/server/Server.java [hunk 9]
```diff
@@ -169,6 +193,10 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
 
         Serializers.INTEGER.write(out, NetworkCode.GET_MESSAGES_BY_ID_RESPONSE);
         Serializers.collection(Message.SERIALIZER).write(out, messages);
+      if (type == NetworkCode.SERVER_INFO_REQUEST) {
```
Comment (aslo) [inline src/codeu/chat/server/Server.java]: See my above comment - you only need to make that change once, so you can probably delete this section.

**Review by aslo — CHANGES_REQUESTED** [review]
Looking good! I left a few comments that you should address or fix before merging this pull request. 

Also, don't forget to compile and run your code before submitting it to make sure it works!

File: src/codeu/chat/client/commandline/Chat.java (hunk 1 above)
Comment (aslo) [reply src/codeu/chat/client/commandline/Chat.java]: Now that you've merged in the changes from master, you can just delete this block of code. Make sure to still add your own code that prints the uptime information to the info command below!

File: src/codeu/chat/client/core/Context.java [hunk 10]
```diff
@@ -47,6 +47,11 @@ public UserContext create(String name) {
       users.add(new UserContext(user, view, controller));
     }
     return users;
+      
+  public Info getInfo() {
```
Comment (aslo) [inline src/codeu/chat/client/core/Context.java]: You can delete this method as well. I don't believe this is valid anymore.

File: src/codeu/chat/client/core/View.java [hunk 11]
```diff
@@ -162,3 +162,19 @@ public ServerInfo getInfo() {
     return null;
 }
 }
+
+  public ServerInfo getInfo() {
```
Comment (aslo) [inline src/codeu/chat/client/core/View.java]: You can delete this section as well. Your merge brought in a duplicated version of this code from master. (See lines 146-164)

File: src/codeu/chat/server/Server.java (hunk 3 above)
Comment (aslo) [reply src/codeu/chat/server/Server.java]: You can delete this.

File: src/codeu/chat/server/Server.java (hunk 4 above)
Comment (aslo) [reply src/codeu/chat/server/Server.java]: You can delete this.

File: src/codeu/chat/server/Server.java (hunk 5 above)
Comment (aslo) [reply src/codeu/chat/server/Server.java]: You can delete this.

File: src/codeu/chat/server/Server.java (hunk 6 above)
Comment (aslo) [reply src/codeu/chat/server/Server.java]: You can delete this.

File: src/codeu/chat/server/Server.java (hunk 7 above)
Comment (aslo) [reply src/codeu/chat/server/Server.java]: You can delete this.

File: src/codeu/chat/server/Server.java (hunk 8 above)
Comment (aslo) [reply src/codeu/chat/server/Server.java]: You can delete this.

File: src/codeu/chat/server/Server.java (hunk 9 above)
Comment (aslo) [reply src/codeu/chat/server/Server.java]: You can delete this.

File: src/codeu/chat/server/Server.java (hunk 2 above)
Comment (aslo) [reply src/codeu/chat/server/Server.java]: You can delete this.

File: src/codeu/chat/common/NetworkCode.java [hunk 12]
```diff
@@ -35,7 +36,8 @@
       RELAY_READ_REQUEST = 27,
       RELAY_READ_RESPONSE = 28,
       RELAY_WRITE_REQUEST = 29,
-      RELAY_WRITE_RESPONSE = 30,
-      SERVER_INFO_REQUEST = 31,
+      RELAY_WRITE_RESPONSE = 30;
+      SERVER_INFO_REQUEST = 31;
```
Comment (aslo) [inline src/codeu/chat/common/NetworkCode.java]: This line, and the line above it, need to end in commas, not semicolons.

File: src/codeu/chat/client/commandline/Chat.java [hunk 13]
```diff
@@ -45,7 +45,7 @@
   private final Stack<Panel> panels = new Stack<>();
 
   public Chat(Context context) {
-    this.panels.push(createRootPanel(context));
```
Comment (aslo) [inline src/codeu/chat/client/commandline/Chat.java]: This line should not have been deleted.

**Review by aslo — CHANGES_REQUESTED** [review]
Very close! You also need to add the line that prints uptime information.

File: src/codeu/chat/client/commandline/Chat.java [hunk 14]
```diff
@@ -147,6 +147,7 @@ public void invoke(List<String> args) {
           } else {
             //Print server info
             System.out.println("Version:" + info.version);
+            System.out.println("UpTime:" + info.upTime);
```
Comment (aslo) [inline src/codeu/chat/client/commandline/Chat.java]: This doens't seem to work. The code doesn't compile because `upTime` doesn't exist as a member of `info`. You'll want to calculate the length of time that the server has been up here.

File: src/codeu/chat/common/ServerInfo.java [hunk 15]
```diff
@@ -5,9 +5,11 @@
 import codeu.chat.util.Uuid;
 public final class ServerInfo {
   private final static String SERVER_VERSION = "1.0.0";
-
+    
+  public final Time startTime;
```
Comment (aslo) [inline src/codeu/chat/common/ServerInfo.java]: This class doesn't compile - I think you still need to import the `Time` class.

**Review by aslo — CHANGES_REQUESTED** [review]
This code still doens't compile - I left some suggestions, but make sure you test it out locally before committing!

File: build.py [hunk 16]
```diff
@@ -44,8 +44,8 @@
   'out' : 'bin',
   'src' : [ 'src', 'test' ],
   'libraries' : [
-    'third_party/junit4-4.11.jar',
-    'third_party/hamcrest-core-1.3.jar'
+    'third_party\junit4-4.11.jar',
```
Comment (aslo) [inline build.py]: You shouldn't commit the changes to this file, since they are specific to windows. Are these changes actually necessary in order to build on windows? Either way, you should undo these changes.

File: src/codeu/chat/server/View.java [hunk 17]
```diff
@@ -77,7 +80,10 @@ public View(Model model) {
 
   @Override
   public Message findMessage(Uuid id) { return model.messageById().first(id); }
-
+    
+  @Override
+  public ServerInfo getInfo() { return info; }
```
Comment (aslo) [inline src/codeu/chat/server/View.java]: I don't think this method is actually called by your code. There's already a `getInfo` method in Context.java that's already used. You should remove this method since it's unused.

File: src/codeu/chat/server/View.java [hunk 18]
```diff
@@ -37,12 +37,15 @@
...
 import codeu.chat.util.store.StoreAccessor;
+import codeu.chat.common.ServerInfo;
 
 public final class View implements BasicView, SinglesView {
 
   private final static Logger.Log LOG = Logger.newLog(View.class);
 
   private final Model model;
+    
+  private static final ServerInfo info = new ServerInfo();
```
Comment (aslo) [inline src/codeu/chat/server/View.java]: You can remove this line - see my comment below in this file.

File: src/codeu/chat/common/BasicView.java [hunk 19]
```diff
@@ -51,5 +52,7 @@
   //
   //   Return all messages whose id is found in the given collection.
   Collection<Message> getMessages(Collection<Uuid> ids);
+    
+  ServerInfo getInfo();
```
Comment (aslo) [inline src/codeu/chat/common/BasicView.java]: I don't think this method is actually called by your code. There's already a getInfo method in Context.java that's already used. You should remove this method since it's unused.

**Review by aslo — CHANGES_REQUESTED** [review]
You shouldn't commit chat_client_log.log.lck or chat_server_log.log.lck, since those are specific to your local workspace only. You should remove them using `git rm chat_client_log.log.lck  chat_server_log.log.lck`

You're really close!!

File: src/codeu/chat/server/View.java [hunk 20]
```diff
@@ -94,7 +95,9 @@ public InterestSet getInterestSet(Uuid id){
 
   @Override
   public Message findMessage(Uuid id) { return model.messageById().first(id); }
-
+    
+  @Override
```
Comment (aslo) [inline src/codeu/chat/server/View.java]: I think you need to delete this annotation as well.

**Review by aslo — APPROVED** [review]
